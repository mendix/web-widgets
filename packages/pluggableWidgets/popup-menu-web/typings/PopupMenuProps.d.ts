/**
 * This file was generated from PopupMenu.xml
 * WARNING: All changes made to this file will be overwritten
 * @author Mendix Widgets Framework Team
 */
import { ComponentType, CSSProperties, ReactNode } from "react";
import { ActionValue, DynamicValue } from "mendix";

export type ItemTypeEnum = "item" | "divider";

export type StyleClassEnum = "defaultStyle" | "inverseStyle" | "primaryStyle" | "infoStyle" | "successStyle" | "warningStyle" | "dangerStyle";

export interface BasicItemsType {
    itemType: ItemTypeEnum;
    caption?: DynamicValue<string>;
    visible?: DynamicValue<boolean>;
    action?: ActionValue;
    styleClass: StyleClassEnum;
}

export interface CustomItemsType {
    content: ReactNode;
    visible?: DynamicValue<boolean>;
    action?: ActionValue;
}

export type TriggerEnum = "onclick" | "onhover";

export type HoverCloseOnEnum = "onClickOutside" | "onHoverLeave";

export type PositionEnum = "left" | "right" | "top" | "bottom";

export type ClippingStrategyEnum = "absolute" | "fixed";

export interface BasicItemsPreviewType {
    itemType: ItemTypeEnum;
    caption: string;
    visible: string;
    action: {} | null;
    styleClass: StyleClassEnum;
}

export interface CustomItemsPreviewType {
    content: { widgetCount: number; renderer: ComponentType<{ children: ReactNode; caption?: string }> };
    visible: string;
    action: {} | null;
}

export interface PopupMenuContainerProps {
    name: string;
    class: string;
    style?: CSSProperties;
    tabIndex?: number;
    advancedMode: boolean;
    menuTrigger: ReactNode;
    basicItems: BasicItemsType[];
    customItems: CustomItemsType[];
    trigger: TriggerEnum;
    hoverCloseOn: HoverCloseOnEnum;
    position: PositionEnum;
    clippingStrategy: ClippingStrategyEnum;
    menuToggle: boolean;
}

export interface PopupMenuPreviewProps {
    /**
     * @deprecated Deprecated since version 9.18.0. Please use class property instead.
     */
    className: string;
    class: string;
    style: string;
    styleObject?: CSSProperties;
    readOnly: boolean;
    renderMode?: "design" | "xray" | "structure";
    advancedMode: boolean;
    menuTrigger: { widgetCount: number; renderer: ComponentType<{ children: ReactNode; caption?: string }> };
    basicItems: BasicItemsPreviewType[];
    customItems: CustomItemsPreviewType[];
    trigger: TriggerEnum;
    hoverCloseOn: HoverCloseOnEnum;
    position: PositionEnum;
    clippingStrategy: ClippingStrategyEnum;
    menuToggle: boolean;
}
