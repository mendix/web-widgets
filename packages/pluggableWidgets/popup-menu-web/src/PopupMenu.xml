<?xml version="1.0" encoding="utf-8" ?>
<widget id="com.mendix.widget.web.popupmenu.PopupMenu" pluginWidget="true" offlineCapable="true" xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../../../../node_modules/mendix/custom_widget.xsd">
    <name>Pop-up menu</name>
    <description>Displays a set of pre-defined items within the Pop-up menu</description>
    <studioProCategory>Menus &amp; navigation</studioProCategory>
    <studioCategory>Menus</studioCategory>
    <helpUrl>https://docs.mendix.com/appstore/widgets/popup-menu</helpUrl>
    <properties>
        <propertyGroup caption="General">
            <propertyGroup caption="General">
                <property key="advancedMode" type="boolean" defaultValue="false">
                    <caption>Enable advanced options</caption>
                    <description />
                </property>
                <property key="menuTrigger" type="widgets" required="false">
                    <caption>The area to open or close the menu.</caption>
                    <description>Responsible for toggling the Pop-up menu.</description>
                </property>
                <property key="basicItems" type="object" isList="true" required="false">
                    <caption>Menu items</caption>
                    <description>The popup menu items.</description>
                    <properties>
                        <property key="itemType" type="enumeration" defaultValue="item">
                            <caption>Item type</caption>
                            <category>General</category>
                            <description />
                            <enumerationValues>
                                <enumerationValue key="item">Button</enumerationValue>
                                <enumerationValue key="divider">Divider</enumerationValue>
                            </enumerationValues>
                        </property>
                        <property key="caption" type="textTemplate" required="false">
                            <caption>Caption</caption>
                            <category>General</category>
                            <description />
                        </property>
                        <property key="visible" type="expression" defaultValue="true" required="false">
                            <caption>Visible</caption>
                            <category>General</category>
                            <description />
                            <returnType type="Boolean" />
                        </property>
                        <property key="action" type="action" required="false">
                            <caption>On click action</caption>
                            <category>General</category>
                            <description />
                        </property>
                        <property key="styleClass" type="enumeration" defaultValue="defaultStyle">
                            <caption>Style</caption>
                            <category>General</category>
                            <description>An extra class will be added: "popupmenu-basic-item-[style]"</description>
                            <enumerationValues>
                                <enumerationValue key="defaultStyle">Default</enumerationValue>
                                <enumerationValue key="inverseStyle">Inverse</enumerationValue>
                                <enumerationValue key="primaryStyle">Primary</enumerationValue>
                                <enumerationValue key="infoStyle">Info</enumerationValue>
                                <enumerationValue key="successStyle">Success</enumerationValue>
                                <enumerationValue key="warningStyle">Warning</enumerationValue>
                                <enumerationValue key="dangerStyle">Danger</enumerationValue>
                            </enumerationValues>
                        </property>
                    </properties>
                </property>
                <property key="customItems" type="object" isList="true" required="false">
                    <caption>Menu items</caption>
                    <description>The popup menu custom items. To make sure the popup closes correctly after a click, do not configure clickable widgets inside the placeholders. Use the action property of this widget.</description>
                    <properties>
                        <property key="content" type="widgets">
                            <caption>Content</caption>
                            <category>General</category>
                            <description />
                        </property>
                        <property key="visible" type="expression" defaultValue="true" required="false">
                            <caption>Visible</caption>
                            <category>General</category>
                            <description />
                            <returnType type="Boolean" />
                        </property>
                        <property key="action" type="action" required="false">
                            <caption>On click action</caption>
                            <category>General</category>
                            <description />
                        </property>
                    </properties>
                </property>
                <property key="trigger" type="enumeration" defaultValue="onclick">
                    <caption>Open on</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="onclick">Click</enumerationValue>
                        <enumerationValue key="onhover">Hover</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="hoverCloseOn" type="enumeration" defaultValue="onHoverLeave">
                    <caption>Close on</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="onClickOutside">Click outside</enumerationValue>
                        <enumerationValue key="onHoverLeave">Hover leave</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="position" type="enumeration" defaultValue="bottom">
                    <caption>Menu position</caption>
                    <description>The location of the menu relative to the click area.</description>
                    <enumerationValues>
                        <enumerationValue key="left">Left</enumerationValue>
                        <enumerationValue key="right">Right</enumerationValue>
                        <enumerationValue key="top">Top</enumerationValue>
                        <enumerationValue key="bottom">Bottom</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="clippingStrategy" type="enumeration" defaultValue="absolute">
                    <caption>Clipping strategy</caption>
                    <description>'Absolute' positions the floating element relative to its nearest positioned ancestor, while 'Fixed' breaks it out of any clipping ancestor.</description>
                    <enumerationValues>
                        <enumerationValue key="absolute">Absolute</enumerationValue>
                        <enumerationValue key="fixed">Fixed</enumerationValue>
                    </enumerationValues>
                </property>
            </propertyGroup>
            <propertyGroup caption="Development">
                <property key="menuToggle" type="boolean" defaultValue="false">
                    <caption>Show preview</caption>
                    <description>Use this to see a preview of the menu items while developing.</description>
                </property>
            </propertyGroup>
        </propertyGroup>
    </properties>
</widget>
