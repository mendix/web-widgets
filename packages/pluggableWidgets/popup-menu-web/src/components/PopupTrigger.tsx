import { useMergeRefs } from "@floating-ui/react";
import { createElement, forwardRef, PropsWithChildren, ReactElement, RefObject } from "react";
import { usePopupContext } from "../hooks/usePopupContext";

export const PopupTrigger = forwardRef(
    ({ children }: PropsWithChildren, propRef: RefObject<HTMLDivElement>): ReactElement => {
        const { getReferenceProps, open, refs } = usePopupContext();
        const childrenRef = (children as any).ref;
        const ref = useMergeRefs([refs.setReference, propRef, childrenRef]);

        return (
            <div
                className={"popupmenu-trigger"}
                ref={ref}
                data-state={open ? "open" : "closed"}
                {...getReferenceProps?.({
                    onClick: e => {
                        e.preventDefault();
                        e.stopPropagation();
                    }
                })}
            >
                {children}
            </div>
        );
    }
);
