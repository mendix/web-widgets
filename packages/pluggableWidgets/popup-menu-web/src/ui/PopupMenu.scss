.popupmenu {
    position: relative;
    display: flex;
    overflow: visible;
}

ul.popupmenu-menu {
    list-style-type: none;
    padding-inline-start: 0;
}

.table .td > .popupmenu {
    overflow: visible;
}

.popupmenu-trigger {
    position: relative;
    cursor: pointer;
}

.popupmenu-menu {
    display: none;
    position: absolute;
    z-index: 999;
    width: max-content;
    flex-direction: column;
    background-color: #ffffff;
    border-radius: 8px;
    box-shadow: 0 2px 20px 1px rgba(5, 15, 129, 0.05), 0 2px 16px 0 rgba(33, 43, 54, 0.08);

    &.popupmenu-position-left {
        top: 0;
        left: 0;
        transform: translateX(-100%);
    }

    &.popupmenu-position-right {
        top: 0;
        right: 0;
        transform: translateX(100%);
    }

    &.popupmenu-position-top {
        top: 0;
        transform: translateY(-100%);
    }

    &.popupmenu-position-bottom {
        bottom: initial;
        top: 0;
        transform: none;
    }

    .popupmenu-basic-item:first-child,
    .popupmenu-custom-item:first-child {
        border-top-left-radius: 8px;
        border-top-right-radius: 8px;
    }

    .popupmenu-basic-item:last-child,
    .popupmenu-custom-item:last-child {
        border-bottom-left-radius: 8px;
        border-bottom-right-radius: 8px;
    }
}

.popupmenu-basic-divider {
    height: 1px;
    width: 100%;
    background-color: #ced0d3;
}

.popupmenu-basic-item {
    color: #555555;
    font-size: 14px;
    padding: 12px 16px;

    &:hover,
    &:focus,
    &:active {
        cursor: pointer;
        border-color: #f5f6f6;
        background-color: #f5f6f6;
    }

    &-inverse {
        color: #252c36;
    }

    &-primary {
        color: #0595db;
    }

    &-info {
        color: #48b0f7;
    }

    &-success {
        color: #76ca02;
    }

    &-warning {
        color: #f99b1d;
    }

    &-danger {
        color: #ed1c24;
    }
}

.popupmenu-custom-item {
    &:hover,
    &:focus,
    &:active {
        cursor: pointer;
        border-color: #f5f6f6;
        background-color: #f5f6f6;
    }
}
