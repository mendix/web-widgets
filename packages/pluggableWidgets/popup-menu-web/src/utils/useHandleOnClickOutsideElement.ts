import { useEffect, RefObject } from "react";

export function useHandleOnClickOutsideElement(ref: RefObject<HTMLUListElement>, handler: () => void): void {
    useEffect(() => {
        const listener = (event: MouseEvent & { target: Node | null }): void => {
            if (!ref.current || ref.current.contains(event.target)) {
                return;
            }
            handler();
        };
        ref.current?.ownerDocument.addEventListener("mousedown", listener);
        ref.current?.ownerDocument.addEventListener("touchstart", listener);
        return () => {
            ref.current?.ownerDocument.removeEventListener("mousedown", listener);
            ref.current?.ownerDocument.removeEventListener("touchstart", listener);
        };
    }, [ref, handler]);
}
