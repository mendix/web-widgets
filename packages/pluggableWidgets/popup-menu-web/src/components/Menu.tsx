import classNames from "classnames";
import { ActionValue } from "mendix";
import { createElement, ReactElement, useEffect, useRef } from "react";
import { createPortal } from "react-dom";
import { BasicItemsType, CustomItemsType, PopupMenuContainerProps, PositionEnum } from "../../typings/PopupMenuProps";
import { useOnClickOutside } from "@mendix/widget-kit-web/hooks/useOnClickOutside";
import { useMenuPlacement } from "../utils/useMenuPlacement";

import {
    isBehindElement,
    isBehindRandomElement,
    isElementPartiallyOffScreen,
    isElementVisibleByUser,
    moveAbsoluteElementOnScreen,
    unBlockAbsoluteElementBottom,
    unBlockAbsoluteElementLeft,
    unBlockAbsoluteElementRight,
    unBlockAbsoluteElementTop
} from "../utils/document";

export interface MenuProps extends PopupMenuContainerProps {
    anchorElement: HTMLDivElement;
    onCloseRequest: () => void;
    onItemClick: (itemAction: ActionValue) => void;
}
export function Menu(props: MenuProps): ReactElement {
    const popupRef = useRef<HTMLUListElement>(null);
    const anchorElement = props.anchorElement;
    const popupStyles = useMenuPlacement(anchorElement, props.position);

    useOnClickOutside(popupRef, props.onCloseRequest);

    useEffect(() => {
        if (popupRef.current) {
            correctPosition(popupRef.current, props.position);
        }
    }, [popupRef.current, props.position]);

    const menuOptions = createMenuOptions(props, props.onItemClick);

    return createPortal(
        <div className="widget-popupmenu-root">
            <ul
                ref={popupRef}
                style={popupStyles}
                className={classNames(
                    "popupmenu-menu",
                    `popupmenu-position-${props.position}`,
                    "popup-portal",
                    props.class
                )}
            >
                {menuOptions}
            </ul>
        </div>,
        anchorElement
    );
}

function checkVisibility(item: BasicItemsType | CustomItemsType): boolean {
    if (Object.prototype.hasOwnProperty.call(item, "visible")) {
        return !!item.visible?.value;
    }
    return true;
}

function createMenuOptions(
    props: PopupMenuContainerProps,
    handleOnClickItem: (itemAction?: ActionValue) => void
): ReactElement[] {
    if (!props.advancedMode) {
        return props.basicItems
            .filter(item => checkVisibility(item))
            .map((item, index) => {
                if (item.itemType === "divider") {
                    return <li key={index} className={"popupmenu-basic-divider"} />;
                } else {
                    const pickedStyle =
                        item.styleClass !== "defaultStyle"
                            ? "popupmenu-basic-item-" + item.styleClass.replace("Style", "")
                            : "";
                    return (
                        <li
                            key={index}
                            className={classNames("popupmenu-basic-item", pickedStyle)}
                            onClick={e => {
                                e.preventDefault();
                                e.stopPropagation();
                                handleOnClickItem(item.action);
                            }}
                        >
                            {item.caption?.value ?? ""}
                        </li>
                    );
                }
            });
    } else {
        return props.customItems
            .filter(item => checkVisibility(item))
            .map((item, index) => (
                <li
                    key={index}
                    className={"popupmenu-custom-item"}
                    onClick={e => {
                        e.preventDefault();
                        e.stopPropagation();
                        handleOnClickItem(item.action);
                    }}
                >
                    {item.content}
                </li>
            ));
    }
}

function correctPosition(element: HTMLElement, position: PositionEnum): void {
    const dynamicDocument: Document = element.ownerDocument;
    const dynamicWindow = dynamicDocument.defaultView as Window;
    let boundingRect: DOMRect = element.getBoundingClientRect();
    const isOffScreen = isElementPartiallyOffScreen(dynamicWindow, boundingRect);
    if (isOffScreen) {
        moveAbsoluteElementOnScreen(dynamicWindow, element, boundingRect);
    }

    boundingRect = element.getBoundingClientRect();
    const blockingElement = isBehindRandomElement(dynamicDocument, element, boundingRect, 3, "popupmenu");
    if (blockingElement && isElementVisibleByUser(dynamicDocument, dynamicWindow, blockingElement)) {
        unBlockAbsoluteElement(element, boundingRect, blockingElement.getBoundingClientRect(), position);
    } else if (blockingElement) {
        let node = blockingElement;
        do {
            if (isBehindElement(element, node, 3) && isElementVisibleByUser(dynamicDocument, dynamicWindow, node)) {
                return unBlockAbsoluteElement(element, boundingRect, node.getBoundingClientRect(), position);
            } else if (node.parentElement) {
                node = node.parentElement as HTMLElement;
            } else {
                break;
            }
        } while (node.parentElement);
    }
}

function unBlockAbsoluteElement(
    element: HTMLElement,
    boundingRect: DOMRect,
    blockingElementRect: DOMRect,
    position: PositionEnum
): void {
    switch (position) {
        case "left":
            unBlockAbsoluteElementLeft(element, boundingRect, blockingElementRect);
            unBlockAbsoluteElementBottom(element, boundingRect, blockingElementRect);
            break;
        case "right":
            unBlockAbsoluteElementRight(element, boundingRect, blockingElementRect);
            unBlockAbsoluteElementBottom(element, boundingRect, blockingElementRect);
            break;
        case "top":
            unBlockAbsoluteElementTop(element, boundingRect, blockingElementRect);
            break;
        case "bottom":
            unBlockAbsoluteElementBottom(element, boundingRect, blockingElementRect);
            break;
    }
}
