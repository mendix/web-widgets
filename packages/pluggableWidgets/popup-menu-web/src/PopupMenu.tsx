import { ReactElement, createElement } from "react";

import { PopupMenuContainerProps } from "../typings/PopupMenuProps";
import { PopupMenu as PopupMenuComponent } from "./components/PopupMenu";

export default function PopupMenu(props: PopupMenuContainerProps): ReactElement {
    return <PopupMenuComponent {...props} />;
}
