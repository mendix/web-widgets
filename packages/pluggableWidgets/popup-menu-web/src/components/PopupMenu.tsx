import { executeAction } from "@mendix/widget-plugin-platform/framework/execute-action";
import classNames from "classnames";
import { ActionValue } from "mendix";
import { createElement, ReactElement, useCallback, useEffect, useState } from "react";
import { PopupMenuContainerProps } from "../../typings/PopupMenuProps";
import { usePopup } from "../hooks/usePopup";
import { Menu } from "./Menu";
import { PopupContext } from "./PopupContext";
import { PopupTrigger } from "./PopupTrigger";

export interface PopupMenuProps extends PopupMenuContainerProps {
    preview?: boolean;
}

export function PopupMenu(props: PopupMenuProps): ReactElement {
    const preview = !!props.preview;
    const [visibility, setVisibility] = useState(preview && props.menuToggle);
    const open = visibility;
    const popup = usePopup({
        open,
        onOpenChange: setVisibility,
        placement: props.position,
        trigger: props.trigger,
        clippingStrategy: props.clippingStrategy
    });

    const handleOnClickItem = useCallback((itemAction?: ActionValue): void => {
        setVisibility(false);
        executeAction(itemAction);
    }, []);

    useEffect(() => {
        setVisibility(props.menuToggle);
    }, [props.menuToggle]);

    return (
        <PopupContext.Provider value={popup}>
            <div className={classNames("popupmenu", props.class)}>
                <PopupTrigger>{props.menuTrigger}</PopupTrigger>
                <Menu {...props} onItemClick={handleOnClickItem} />
            </div>
        </PopupContext.Provider>
    );
}
