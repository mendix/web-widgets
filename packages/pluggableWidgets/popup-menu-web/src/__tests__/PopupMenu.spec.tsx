import { render, RenderResult } from "@testing-library/react";
import { ValueStatus } from "mendix";
import { createElement } from "react";
import { BasicItemsType, CustomItemsType, PopupMenuContainerProps } from "../../typings/PopupMenuProps";
import { PopupMenu } from "../components/PopupMenu";
import { dynamicValue } from "../utils/attrValue";

import "@testing-library/jest-dom";

jest.useFakeTimers();

describe("Popup Menu", () => {
    const createPopupMenu = (props: PopupMenuContainerProps): RenderResult => render(<PopupMenu {...props} />);
    const basicItemProps: BasicItemsType = {
        itemType: "item",
        caption: dynamicValue("Caption"),
        styleClass: "defaultStyle"
    };
    const customItemProps: CustomItemsType = { content: createElement("div", null, null) };

    const defaultProps: PopupMenuContainerProps = {
        name: "popup-menu",
        class: "mx-popup-menu",
        tabIndex: -1,
        trigger: "onclick",
        menuToggle: true,
        menuTrigger: createElement("button", null, "Trigger"),
        advancedMode: false,
        position: "bottom",
        hoverCloseOn: "onHoverLeave",
        basicItems: [
            basicItemProps,
            { itemType: "divider", caption: dynamicValue("Caption"), styleClass: "defaultStyle" }
        ],
        customItems: [customItemProps],
        clippingStrategy: "absolute"
    };

    it("renders popup menu", () => {
        const popupMenu = createPopupMenu(defaultProps);

        expect(popupMenu.asFragment()).toMatchSnapshot();
    });

    describe("with custom items", () => {
        beforeEach(() => {
            defaultProps.advancedMode = true;
        });

        it("renders custom items without hidden items", () => {
            const customItem: CustomItemsType = {
                ...customItemProps,
                visible: {
                    value: false,
                    status: ValueStatus.Available
                }
            };
            const { container } = createPopupMenu({
                ...defaultProps,
                customItems: [customItem]
            });
            expect(container.querySelectorAll(".popupmenu-custom-item")).toHaveLength(0);
        });
    });
});
