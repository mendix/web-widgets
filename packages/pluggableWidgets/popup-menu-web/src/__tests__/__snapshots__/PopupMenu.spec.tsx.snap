// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup Menu renders popup menu 1`] = `
<DocumentFragment>
  <div
    class="popupmenu mx-popup-menu"
  >
    <div
      aria-controls=":r0:"
      aria-expanded="true"
      aria-haspopup="dialog"
      aria-hidden="true"
      class="popupmenu-trigger"
      data-floating-ui-inert=""
      data-state="open"
    >
      <button>
        Trigger
      </button>
    </div>
    <span
      aria-hidden="true"
      data-floating-ui-focus-guard=""
      data-floating-ui-inert=""
      data-type="inside"
      role="button"
      style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: fixed; white-space: nowrap; width: 1px; top: 0px; left: 0px;"
      tabindex="0"
    />
    <div
      class="widget-popupmenu-root"
    >
      <ul
        class="popupmenu-menu"
        data-floating-ui-focusable=""
        id=":r0:"
        role="dialog"
        style="position: absolute; left: 0px; top: 0px; transform: translate(0px, 0px);"
        tabindex="0"
      >
        <li
          class="popupmenu-basic-item"
        >
          Caption
        </li>
        <li
          class="popupmenu-basic-divider"
        />
      </ul>
    </div>
    <span
      aria-hidden="true"
      data-floating-ui-focus-guard=""
      data-floating-ui-inert=""
      data-type="inside"
      role="button"
      style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: fixed; white-space: nowrap; width: 1px; top: 0px; left: 0px;"
      tabindex="0"
    />
  </div>
</DocumentFragment>
`;
